### ISSUE TITLE
Built in TCP/UDP (UNIX stream/datagram) logger handler

### ISSUE THREAD

**Issue by hauleth** [erlang/otp#5561]
**Is your feature request related to a problem? Please describe.**

A lot of centralised logging subsystems that gather logs locally works over the UNIX datagram sockets or UDP sockets, examples of such subsystems:

- `journald`
- Different `syslog` implementations (`rsyslog`, `syslog-ng`)
- [Vector](https://vector.dev/docs/reference/configuration/sources/socket/)
- Many more systems that I am not aware of

Implementing such thing "on your own" is problematic with current state of handlers as there is no way to have overload protection without additional libraries (like mine [`enough`](https://github.com/hauleth/enough) that is almost direct extraction of `logger_olp` with some private `logger` API removed)

**Describe the solution you'd like**

There should be `logger_socket_h` or `logger_udp_h`/`logger_tcp_h` modules that would provide logging over socket with proper overload protection.

**Describe alternatives you've considered**

Implementing them as external library, but due to #5560 and lack of "official" way to overload problem I think it would be better to have it as part of the `kernel`. Especially as it may be useful in almost all systems with proper formatters or can be used as a base for other handlers that would wrap this functionality with custom additions.

**Comment by garazdawi** [comment]
To have socket handlers in general I think the better approach is to solve the problem by either exposing logger_olp (as you suggested in #4050), or possibly adding implementing a `file_server` process that wraps `socket` so that you can use `logger_std_h` and just set the socket process as the [`device` type](https://www.erlang.org/doc/man/logger_std_h.html#description).

I'm unsure why the udp backend would need any overload protection mechanism? Isn't overflow protection built into the protocol? If I were to build a UDP logger I would just call `gen_udp:send` in the logging process directly and let UDP drop any data that cannot be delivered to the receiving server.

**Comment by hauleth** [comment]
@garazdawi TCP has congestion protection, UDP does not have any of such things AFAIK. So in case of high throughput it could be problematic. However now that I think it would be easier to create one socket per scheduler and then just make the handler to pick one for the current scheduler. This should make the "overload" negligible.

With `file_server` the question is whether the `logger_std_h` do anything to the multiline messages, because the big thing about using packeted transport (like UDP) is that you can easily send multiline messages as parts of the same package, and then you will end with single, multiline, log entry in contrast to stream protocols like TCP or stdout/stderr.

**Comment by garazdawi** [comment]
> With file_server the question is whether the logger_std_h do anything to the multiline messages, because the big thing about using packeted transport (like UDP) is that you can easily send multiline messages as parts of the same package, and then you will end with single, multiline, log entry in contrast to stream protocols like TCP or stdout/stderr.

The current implementation will send one `io_request` with the entire log message to the device, for right now it will work like that. Since the olp is built to load balance on per log message, I do not see any reason for that to change, so I think you can rely on that. If you want to make sure that you can rely on it, we would welcome a testcase that tests that it is done that way.

**Comment by seriyps** [comment]
Implementing a file server protocol would mean log message would have to travel through two processes then?

Original process -> logger_std_h process -> Process implementing `file_server` -> gen_tcp:send/gen_udp:send

But I also think that it would be better to extract some generic overload protection library from logger rather than implementing tons of ad-hoc handlers.
I ended up implementing my own partial implementation of [overload protection for journald logger handler](https://github.com/seriyps/logger_journald/blob/34b5d9cc0c35b06abe05fe2b5b7d826bf69fd8bc/src/logger_journald_h.erl#L347-L420) and it was indeed the most complex part of handler implementation.

**Comment by hauleth** [comment]
@seriyps FYI I also have implementation of [journald handler](https://github.com/hauleth/erlang-systemd/blob/master/src/systemd_journal_h.erl). I am still polishing the library, but most of the features are complete.

**Comment by seriyps** [comment]
@hauleth yes, I remember we discussed it a bit over email about a year ago :+1:

**Comment by garazdawi** [comment]
> Implementing a file server protocol would mean log message would have to travel through two processes then?
> 
> Original process -> logger_std_h process -> Process implementing `file_server` -> gen_tcp:send/gen_udp:send

There is actually yet another process in-between "logger_std_h process" and "Process implementing `file_server`" that is there to make sure that any blocking call to io:put_chars or file:write does not block "logger_std_h process". Those calls can block if you write to stdout or a file on an nfs, and if the "logger_std_h process" gets blocked it may not be able to load shed in time if a spike in log messages happens.

Since the API to write to an I/O device is asynchronous, it could be possible for logger_std_h to message the io device directly, but that is not done today.

> And also the receiver of the io_request would only be able to see the already serialized iodata rather than raw logger:log_event() structure, which might be beneficial in many cases.

The built-in logger handlers all do the formatting in the original process as that means that formatting messages can be parallelized. It also makes sure that the data sent from the original process is relatively small (or off-heap).

**Comment by seriyps** [comment]
> The built-in logger handlers all do the formatting in the original process as that means that formatting messages can be parallelized. It also makes sure that the data sent from the original process is relatively small (or off-heap).

Yep, I can still see a potential benefit in custom logger process being able to get the access to the log's metadata, even if it does most of the CPU-intensive preparations outside of the underlying process. I can imagine some logger that logs via HTTP or AMQP and althrough it sends most of the `logger:log_event()` as JSON (formatted straight in the `log/2` callback), it may want to use, say, `meta.domain` or `meta.correlation_id` or `level` or whatever from inside the logger handler process to add some HTTP/AMQP headers. The only way to achieve that, as I see it, is to send not a plain formatted text to this process, but some structure that holds pre-formatted text + some limited subset of the metadata. For my HTTP/AMQP example something like `{log, get([level], Msg), get([meta, domain], Msg), get([meta, correlation_id], Msg), convert_to_json(Msg)}`.
This can't be achieved with `logger_std_h` + `file_server`, but can be achieved with custom logger handler callback module. But then overload protection have to be implemented as well, while it might be cool to have some generic building blocks for overload protection available in OTP I think.

**Comment by hauleth** [comment]
I am working on UDP logger handler - https://github.com/hauleth/logger_udp. This one will create UDP socket per scheduler and then will dispatch the messages depending on current scheduler ID. This way I do not need to care about congestion and overload protection as it is handled "automatically" in OS.
